### PR TITLE
starting on an email serverless service

### PR DESCRIPTION
**Pull request by husobee**
### Summary

Resolves: https://github.com/brave-intl/internal/issues/331

### Type of change ( select one )

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: serverless/email/unsubscribe/main.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	config, err := appaws.BaseAWSConfig(ctx, logger)
+	if err != nil {
+		// failed to get the base aws config
+		return events.APIGatewayProxyResponse{
+			StatusCode: http.StatusInternalServerError,
+			Body:       http.StatusText(http.StatusInternalServerError),
+		}, fmt.Errorf("failed to get the base aws config: %w", err)
+	}
+	// setup dynamodb client
+	dynamoClient := dynamodb.NewFromConfig(config)
```
Comment (nvonpentz) [inline serverless/email/unsubscribe/main.go:35]: I imagine there will be reasonable overhead associated with establishing a new dynamo connection in each handle request.  Perhaps something to address in future iterations

Comment (husobee) [reply serverless/email/unsubscribe/main.go]: resolved here https://github.com/brave-intl/bat-go/pull/1593/commits/814bbf32307bb940831c328eef1defd39a8bd1a6

**Comment by thypon** [comment]
```
  services/promotion/service.go
     trailofbits.go.questionable-assignment.questionable-assignment
        Should `service` be modified when an error could be returned?
        Details: https://sg.run/qq6y

        153┆ service.kafkaWriter, service.kafkaDialer, err = kafkautils.InitKafkaWriter(ctx, suggestionTopic)
          ⋮┆----------------------------------------
        165┆ service.kafkaAdminAttestationReader, err = kafkautils.NewKafkaReader(ctx, groupID, adminAttestationTopic)
          ⋮┆----------------------------------------
        171┆ service.codecs, err = kafkautils.GenerateCodecs(map[string]string{
        172┆ 	"suggestion":          suggestionEventSchema,
        173┆ 	adminAttestationTopic: adminAttestationEventSchema,
        174┆ })
          ⋮┆----------------------------------------
        210┆ service.hotWallet, err = uphold.New(ctx, info, privKey, pubKey)
```

We should probably assign, only if err is nil, here

**Comment by thypon** [comment]
Same here:

```
  services/promotion/controllers.go
     trailofbits.go.questionable-assignment.questionable-assignment
        Should `resp` be modified when an error could be returned?
        Details: https://sg.run/qq6y

        536┆ resp.DrainID, err = service.Drain(ctx, req.Credentials, req.WalletID)
```

File: docker-compose.dev-refresh.yml [hunk 2]
```diff
@@ -12,7 +12,7 @@ services:
     ports:
       - "3335:3333"
       - "6061:6061"
-    command: "go run main.go serve grant"
+    command: "cd main && go run main.go serve grant"
```
Comment (thypon) [inline docker-compose.dev-refresh.yml:15]: Missing `no-new-privileges:true` in `security_opt`, and `read_only: true`

File: serverless/email/webhook/input.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	Supported resource types:
+		DEPOSIT_CRYPTO
+		SUPPORT_TICKET
+		KYC_LEVEL
+		DEPOSIT_FIAT
+		ACCOUNT_FROZEN
+*/
+
+type emailPayload struct {
+	Email        string                 `json:"email"`
```
Comment (thypon) [inline serverless/email/webhook/input.go:17]: Where are we validating `email` and all the other fields to be correctly formatted?

**Review by thypon — COMMENTED** [review]
Are we verifying a webhook signature is there any authentication method for this email webhook?

**Comment by husobee** [comment]
> (1) Are we verifying a webhook signature? (2) Is there any authentication method for this email webhook?

Yes, here is where we are performing an authentication check: https://github.com/brave-intl/bat-go/pull/1593/files#diff-e2d0061654d47f64fcce6785bf82d2b97220bf1e5bde20eb32cbaee2ce7955e1R81-R102

We match the x-api-key header to any of the valid tokens we are storing in secrets manager.

**Comment by husobee** [comment]
fixed questionable assignments here: https://github.com/brave-intl/bat-go/pull/1593/commits/98c316b054b93ec9e115d3f6a5079e4b126606ec

fixed docker-compose files to include security_opts here: https://github.com/brave-intl/bat-go/pull/1593/commits/bb4d7b2dcf9a48fd997fc662a2eb90c2b46d9ae7

File: serverless/email/webhook/input.go (hunk 3 above)
Comment (husobee) [reply serverless/email/webhook/input.go]: addressed here: https://github.com/brave-intl/bat-go/pull/1593/commits/70265af51d6318fdd53c2d07899ff7b5bb969512

File: docker-compose.dev-refresh.yml (hunk 2 above)
Comment (husobee) [reply docker-compose.dev-refresh.yml:18]: addressed here: https://github.com/brave-intl/bat-go/pull/1593/commits/bb4d7b2dcf9a48fd997fc662a2eb90c2b46d9ae7

File: services/wallet/docker-compose.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@ services:
     ports:
       - "3353:3353"
       - "6061:6061"
```
[the remark this answers is not in the document]
Comment (husobee) [reply services/wallet/docker-compose.yml:18]: addressed here: https://github.com/brave-intl/bat-go/pull/1593/commits/bb4d7b2dcf9a48fd997fc662a2eb90c2b46d9ae7

File: Dockerfile [hunk 5]
```diff
@@ -12,13 +12,13 @@ ARG VERSION
...
+RUN cd main && go mod download
 
-RUN CGO_ENABLED=0 GOOS=linux go build \
+RUN cd main && CGO_ENABLED=0 GOOS=linux go build \
     -ldflags "-w -s -X main.version=${VERSION} -X main.buildTime=${BUILD_TIME} -X main.commit=${COMMIT}" \
-    -o bat-go main.go
+    -o ../bat-go main.go
 
 FROM alpine:3.15 as base
 # put certs in artifact from builder
```
[the remark this answers is not in the document]
Comment (husobee) [reply Dockerfile:24]: addressed here: https://github.com/brave-intl/bat-go/pull/1593/commits/bb2a0b3174cb8e905543258d32449197bfeef5e9

[the remark this answers is not in the document]
Comment (thypon) [reply Dockerfile:24]: Shouldn't this be something like `USER nouser` or `USER nobody` depending on the underlying distribution?

[the remark this answers is not in the document]
Comment (husobee) [reply Dockerfile:31]: i understand now.  addressed here: https://github.com/brave-intl/bat-go/pull/1593/commits/b04e90372fcfe02cc06813ee0a65099a1e1b654d

File: docker-compose.yml [hunk 6]
```diff
@@ -90,7 +90,6 @@ services:
       - "6061:6061"
     security_opt:
       - no-new-privileges:true
-      - read_only:true
```
Comment (thypon) [inline docker-compose.yml]: Is this `read_only: true` breaking the appliance?
Ideally, this should work by mounting a dir in a volume/tmpfs with `rw` and `noexec` privs.

Comment (husobee) [reply docker-compose.yml]: yes, unfortunately this particular docker-compose service container is performing a build and this read_only filesystem is breaking it.  FWIW this is only used for CI/local development.

**Review by thypon — DISMISSED** [review]

**Review by clD11 — CHANGES_REQUESTED** [review]

File: libs/custodian/regions.go [hunk 7]
```diff
@@ -21,7 +21,7 @@ var (
 )
 
 // ExtractCustodianRegions - extract the custodian regions from the client
-func ExtractCustodianRegions(ctx context.Context, client appaws.S3GetObjectAPI, bucket string) (*CustodianRegions, error) {
+func ExtractCustodianRegions(ctx context.Context, client appaws.S3GetObjectAPI, bucket string) (*Regions, error) {
```
[the remark this answers is not in the document]
Comment (husobee) [reply libs/custodian/regions.go:24]: https://github.com/brave-intl/bat-go/issues/1628 will handle with this ticket.

File: go.work [hunk 8]
```diff
@@ -1,6 +0,0 @@
-go 1.18
```
[the remark this answers is not in the document]
Comment (husobee) [reply go.work:1]: https://go.googlesource.com/proposal/+/master/design/45713-workspace.md#preventing-files-from-being-checked-in-to-repositories

Seems like go workspaces should not be checked into repository?

File: serverless/email/unsubscribe/main.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	appaws "github.com/brave-intl/bat-go/libs/aws"
+	"github.com/brave-intl/bat-go/libs/logging"
+	"github.com/rs/zerolog"
+)
+
+var (
+	ctx, logger     = logging.SetupLoggerWithLevel(context.Background(), zerolog.InfoLevel)
```
[the remark this answers is not in the document]
Comment (husobee) [reply serverless/email/unsubscribe/main.go]: addressed: https://github.com/brave-intl/bat-go/pull/1593/commits/330c3cba7cc832f7a5fb66e956c2b45553ac0ede

File: serverless/email/status/main.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	appaws "github.com/brave-intl/bat-go/libs/aws"
+	"github.com/brave-intl/bat-go/libs/logging"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+)
+
+var (
+	// setup context/logger
+	ctx, logger = logging.SetupLoggerWithLevel(context.Background(), zerolog.InfoLevel)
```
[the remark this answers is not in the document]
Comment (husobee) [reply serverless/email/status/main.go:22]: addressed: https://github.com/brave-intl/bat-go/pull/1593/commits/330c3cba7cc832f7a5fb66e956c2b45553ac0ede

**Review by thypon — DISMISSED** [review]

File: serverless/email/status/main.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+		case "Open":
+			continue // Skip "Open" events
+		case "Click":
+			continue // Skip "Click" events
+		case "Rendering Failure":
+		case "DeliveryDelay":
+			statusTimestamp = notification.DeliveryDelay.Timestamp
+		case "Subscription":
+		default:
+			logger.Warn().Msg("unknown event type " + notification.EventType)
```
[the remark this answers is not in the document]
Comment (nvonpentz) [reply serverless/email/status/main.go:107]: I opted to not include the StatusTs if it is null https://github.com/brave-intl/bat-go/pull/1636

File: serverless/email/status/main.go [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+		}
+
+		// Get Idempotency key from tags to use as partition key, skip if it is not present
+		var idempotencyKey string
+		for _, tag := range notification.Mail.Tags {
+			idempotencyKey, _ = tag["idempotencyKey"]
```
[the remark this answers is not in the document]
Comment (nvonpentz) [reply serverless/email/status/main.go:113]: Added break https://github.com/brave-intl/bat-go/pull/1636

**Pull request closed without merging**